### PR TITLE
CEL Policy Readme

### LINKED ISSUE #976: [Policy compiler] Document language reference
The policy compiler has a few early adopters, but without a language reference documentation, the syntax has to be guessed from the tests and the code.

### PR DESCRIPTION
**Pull request by TristonianJones**
Initial policy language reference docs.

Closes #976

### CODE REVIEW

File: policy/testdata/README.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# CEL Policy
+
+The Common Expression Language (CEL) supports simple expressions: no variables, functions, or modules. However, CEL expression graphs can be composed together, allowing for reuse and development clarity which is not otherwise possible within CEL.
+
+To address this case, we're introducing the CEL Policy format which is fully runtime compatible with CEL. All of the same performance and safety hardening guarantees which apply to CEL also apply to CEL Policy. The net effect is significantly improved authoring and testability. The YAML-based policy format is easily extensible and inspired by [Kubernetes Admission Policy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) with CEL.
+
+## Policy Language
```
Comment (jcking) [inline policy/testdata/README.md:16]: Originally the top level rule was plural, yeah? Is it always single now?

File: policy/testdata/README.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+A policy is a named instance of a rule which consists of a set of conditional outputs and conditional sub-rules. Matches within the rule and subrules are combined and ordered according to the policy evaluation semantic. The default semantic is `FIRST_MATCH`. The supported top-level fields in a policy include: `name`, `description`, `imports`, and `rule.`
+
+### Rule
+
+The `rule` node in a policy is the primary entry point to CEL computations. Fields above the `rule` are intended to simplify or support the CEL expressions within the `rule` block. For example, the [`imports` list refers to a set of type names](#imports) who should be imported by their simple name within the CEL expressions contained in the `rule`.
+
+#### Variables
+
+A \`rule\` has a single `variables` block. Variables are written as an ordered list. Variables may refer to another variable; however, the  variable must be declared before use, i.e. be defined before it is referenced. A variable has a `name` and an `expression`.
```
Comment (jcking) [inline policy/testdata/README.md]: Backslash around rule? Probably remove them.

Comment (jcking) [inline policy/testdata/README.md]: Double spaces around `the` .

File: policy/testdata/README.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+#### Match
+
+A `rule` has a single `match` block. The match block should have at least one `output` value, though `output` expressions may be `condition`al. The default evaluation order for the sequence of matches is top-down, first-match.
+
+```
+rule:
+  match:
+    - condition: "request.user.name.startsWith('j')"
+      output: "Hi, J!"
+    - output: "Hi, " + request.user.name + "!"
```
Comment (jcking) [inline policy/testdata/README.md]: If the `condition` is missing, what does it mean?

File: policy/testdata/README.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    - condition: "variables.name.startsWith('j')"
+      output: "Hi, J!"
+    - output: "Hi, " + variables.name + "!"
+```
+
+#### Condition
+
+A `condition` expression must type-check to a `bool` return type. When a `condition` predicate evaluates to `true`, either an `output` expression is returned or a nested `rule` result is returned. Using a `condition` with nested `rule` values allows for the declaration of `rule` blocks with local `variables` and reduces the complexity of `condition` expressions within the nested `rule`.
+
+If all \`output\` expressions within a `rule` have associated `condition` predicates, then the return type of the policy is `optional_type(type(output))`. In other words, if the policy is evaluating `true` or `false` output expressions, but all output values are conditional, then the output type of the policy is `optional_type(bool)`. If the nested `rule` does not result in an output, then the `optional.none()` value is returned as the overall policy result.
```
Comment (jcking) [inline policy/testdata/README.md]: Backslashes again.

File: policy/testdata/README.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+Taking our example from earlier, since the `match` is exhaustive and includes a default `output`, then the result type of this policy is `string:`
```
Comment (jcking) [inline policy/testdata/README.md]: `string`, not `string:`

File: policy/testdata/README.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+If we remove the last output, then the result type is `optional_type(string)` since not all evaluation paths will result in an `output`.
+
+```
+rule:
+  match:
+    - condition: "request.user.name.startsWith('j')"
+      output: "Hi, J!"
+```
+
+For more information on optionals, see [https://github.com/google/cel-spec/wiki/proposal-246](https://github.com/google/cel-spec/wiki/proposal-246) for more information about `optional` values within CEL.
```
Comment (jcking) [inline policy/testdata/README.md]: remove the test after the hyperlink?

File: policy/testdata/README.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+#### Output
+
+The `output` field is optional and is, effectively, just like any other CEL expression; however, the output expression types must all agree within the policy expression graph. An output expression may be simple, such as a `bool` or `string` value, or it may be much more complex such as a JSON-like `map` or a structured protocol buffer message.
```
Comment (jcking) [inline policy/testdata/README.md]: Remove structured from before protocol buffer message, a bit repetative.

File: policy/testdata/README.md (hunk 1 above)
Comment (TristonianJones) [reply policy/testdata/README.md:16]: It's been single ever since we introduced nesting support.

File: policy/testdata/README.md (hunk 3 above)
Comment (TristonianJones) [reply policy/testdata/README.md]: I updated the text right after this example. The `condition` defaults to `true` if not specified.

**Comment by TristonianJones** [comment]
PTAL

**Review by jcking — APPROVED** [review]

**Pull request merged**